bm-timeline/src/Resources.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main
+  (
+    main
+  )
+where
+
+import Control.Monad (foldM)
+import Data.List (sortOn)
+import Data.Text.Lazy (Text, pack)
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.IO as TLIO
+import Data.Time.Clock (diffUTCTime)
+import System.Environment (getArgs)
+import System.IO (IOMode(..), withFile)
+import Text.Read (readMaybe)
+
+import Cardano.BM.Common
+import Cardano.BM.Csv
+import qualified Cardano.BM.CPUticks as CPUticks
+import qualified Cardano.BM.MemResident as MemResident
+
+
+{- main procedure -}
+main :: IO ()
+main = do
+    (n, basep) <- parseArguments
+    let nodes = mknodes n
+    process basep nodes
+
+mknodes :: Int -> [NodeId]
+mknodes n = [0 .. n-1]
+
+parseArguments :: IO (Int, FilePath)
+parseArguments = do
+    parsed <- getArgs >>= return . parseArgs
+    case parsed of
+        Left m -> TLIO.putStrLn "call: bmresources <basepath>" >> error m
+        Right res -> return res
+
+  where
+    parseArgs [] = Left "no arguments"
+    parseArgs (_a : []) = Left "not enough arguments"
+    parseArgs (n : bp : []) =
+        case readMaybe n of
+            Nothing -> Left $ "can't parse: " ++ n
+            Just v -> if v > 0
+              then Right (v, bp)
+              else Left $ "wrong value: " ++ n
+    parseArgs _ = Left "too many arguments"
+
+process :: FilePath -> [NodeId] -> IO ()
+process basep nodes = do
+  cputicks0 <- foldM (loadCPUticks basep) [] nodes
+  memusage0 <- foldM (loadMemory basep) [] nodes
+  withFile "resources.csv" WriteMode $ \h -> do
+      let memusage = sortOn fst $ reverse $ map (\(t0,d) -> (TL.toStrict t0, d)) $ processMemory (length nodes) memusage0
+      let cpuusage = sortOn fst $ reverse $ map (\(t0,d) -> (TL.toStrict t0, d)) $ processCPUticks (length nodes) cputicks0
+      -- TLIO.putStrLn . TL.pack . show $ take 20 cpuusage
+      -- TLIO.putStrLn . TL.pack . show $ take 20 memusage
+      let header nm = "timestamp" : (map (\nid -> nm <> " " <> T.pack (show nid)) nodes)
+      output_csv h $ named_columns [ (header "cpu", timestamp_with_list(cpuusage))
+                                   , (header "mem", timestamp_with_list(memusage)) ]
+  return ()
+
+processMemory :: Int -> [MemResident.MemResident] -> [(Text, [Double])]
```

Comment:
Maybe we could introduce `newtype` wrappers for those `Double`'s to highlight their intent?

Reply:
yes, right. and in the future we may need wrap it in a Maybe so we can create "empty" cells.